torch/csrc/api/include/torch/optim/serialize.h

```diff
@@ -207,27 +207,77 @@ void serialize(
   }
 }
 
+template <typename T>
+c10::List<T> deque_to_list(const std::deque<T>& dq) {
+  c10::List<Tensor> list;
+  list.reserve(dq.size());
+  for (const auto& e : dq) {
+    list.push_back(e);
+  }
+  return list;
+}
+
+template <typename T>
+void list_to_deque(const c10::List<T>& list, std::deque<T>& dq) {
+  for (size_t i = 0; i < list.size(); i++) {
+    dq.push_back(list[i]);
+  }
+}
+
 #define _TORCH_OPTIM_SERIALIZE(name) \
   torch::optim::serialize(archive, #name, self.name)
 
 #define _TORCH_OPTIM_SERIALIZE_WITH_TEMPLATE_ARG(OptimizerName) \
   torch::optim::serialize<OptimizerName##ParamState, OptimizerName##Options>(archive, self)
 
-#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name) \
-{ \
+#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name) { \
   auto ivalue = torch::IValue(name()); \
   /* do not serialize if name is an undefined tensor*/ \
   if (!(ivalue.isTensor() && ivalue.unsafeToTensorImpl() == at::UndefinedTensorImpl::singleton())) { \
     archive.write(#name, ivalue); \
   } \
 }
 
+#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG_DEQUE(name) {\
+  if (name().size() != 0) { \
+    c10::IValue ivalue = torch::IValue(deque_to_list(name())); \
+    archive.write(#name, ivalue); \
+  } \
+}
+
 #define _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(T, name) \
 { \
   c10::IValue ivalue; \
   bool exists = archive.try_read(#name, ivalue); \
   if (exists) \
     name(ivalue.to<T>()); \
+  /* undefined tensors are not serialized. it's imp to save {} in name() to ensure that
+  undefined tensor values are retained in case the user set it. */ \
+  else if (ivalue.isTensor()) \
+    name({}); \
 }
+
+#define _TORCH_OPTIM_DESERIALIZE_TORCH_ARG_OPTIONAL(T, name) \
+{ \
+  c10::IValue ivalue; \
+  archive.read(#name, ivalue); \
+  /*c10::nullopt is serialized which is consistent with the
+   Python API as it serializes None values. currently only options (and not state)
+   consist of c10::optional entries so we don't end up serializing anything that's not
+   serialized by Python API*/ \
+  name(ivalue.toOptional<T>()); \
```

Comment:
Would it work if we merge `_TORCH_OPTIM_DESERIALIZE_TORCH_ARG` and this macro? I believe `ivalue.to<T>())` also works when `T` is an `c10::optional` type, and it is worth checking

Reply:
ivalue.to<T>()) doesn't for c10::nullopt :/